### DISCUSSION TITLE
Bridging Effects/Actions to .refreshable(action: @escaping () async -> Void)

### DISCUSSION THREAD

**Discussion by kamcma** [pointfreeco/swift-composable-architecture#648] [Ideas]
Hello!

I know in some other discussions Brandon and Stephen have said they're exploring how TCA might incorporate the new Swift concurrency features directly, and I look forward to it.

In the meantime, though, I've been pondering if there might be a way to bridge the world of Effects and Actions and some of the new async SwiftUI APIs. I'm thinking in particular of .refreshable(action: @escaping () async -> Void).

Consider the basic case of an action that initiates a network request mapped to an action that receives the response. Is there some way to capture that send-and-receive pair of actions such that we could pass a function of the appropriate async signature to that refreshable(action:) modifier? Perhaps implementable as an extension on ViewStore (I'm not proposing the library itself implement any of these stop-gap bridges). I feel like there's something there but I'm not quite sharp enough to put the pieces together myself.

Appreciate anyone's thoughts. Have a nice day.

**Comment by kamcma** [comment]
(More specifically, I see there was a comment that there will be upcoming episodes addressing WWDC topics, so if the answer is "sit tight", that's fine too 😄.)

**Comment by mbrandonw** [comment]
Hi @kamcma, you are right that we will be addressing this soon. We are playing around with some APIs to demonstrate what is possible, and in fact your idea to implement this as an extension on `ViewStore` is how we are approaching it. Something akin to a new `.send` method on `ViewStore` that is `async` so that you can suspend until some state changes signaling that the work is done.
